src/praisonai-bot/praisonai_bot/bots/_outbox.py

```
@@ -477,17 +514,47 @@ def _mark_permanent_failure(self, entry_id: int, error: str) -> None:
+                query = select_cols + """
+                    FROM outbound_queue q
+                    JOIN (
+                        SELECT COALESCE(lane_key, target) AS lane, MIN(id) AS head_id
+                        FROM outbound_queue
+                        WHERE status IN ('pending', 'recovered', 'failed', 'sending')
+                        GROUP BY COALESCE(lane_key, target)
+                    ) h ON COALESCE(q.lane_key, q.target) = h.lane AND q.id = h.head_id
+                    WHERE q.status IN ('pending', 'recovered', 'failed')
+                       OR (q.status = 'sending' AND (q.last_attempt IS NULL OR q.last_attempt <= ?))
+                    ORDER BY q.ts ASC, q.id ASC
+                """
```

Ruff (0.15.20)

[error] Possible SQL injection vector through string-based query construction

(S608)